File: src/create.py
from core import write_session, session_exists
from core_bookmarks import get_session_from_bookmarks

def create(args) -> None:
    _default_mode(args)

def _bookmarks_mode(args) -> None:
    profile_path = args.bookmarks[0]
    bookmarks_path = args.bookmarks[1].lower()
    session = get_session_from_bookmarks(profile_path, bookmarks_path)
    write_session(session, "bookmarks")

def _default_mode(args) -> None:
    name = _get_session_name()
    urls = _get_urls()
    if urls:
        _confirm_creation({
            "name": name,
            "urls": urls
        }) 

def _get_session_name() -> str:
    name = input("Session name: ")
    if session_exists(name, "user"):
        raise RuntimeError(f"{name} session already exists.")
    return name

def _get_urls() -> list:
    urls = []
    while True:
        url = input(f"URL {len(urls) + 1} (q to quit): ")
        if url.lower() != "q":
            urls.append(url)
        else:
            return urls

def _confirm_creation(session: dict) -> None:
    name = session["name"]
    urls = session["urls"]
    confirm = input(f"Confirm creation of {name} (size: {len(urls)}) ? (Y/n) ")
    if confirm.lower() != "n":
        write_session(session, "user")
        print(f"New session 'user/{name}.json' created.")
    else:
        print("Creation aborted. Bye!")

File: src/create.rs
use std::fs::File;
use std::io::{self, Write};
use std::fs::File;
use std::fs::create_dir_all;
use std::path::Path;
use serde_json::json;
use serde_json::to_string_pretty;

pub fn get_input() -> io::Result<(String, Vec<String>)> {
    let mut session_name = String::new();
    let mut urls = Vec::new();

    print!("Session name: ");
    io::stdout().flush()?;
    io::stdin().read_line(&mut session_name)?;

    loop {
        let mut url_input = String::new();
        print!("URL {} (q to quit): ", urls.len() + 1);
        io::stdout().flush()?;
        io::stdin().read_line(&mut url_input)?;
        match url_input.trim() {
            "q" => break,
            value => urls.push(value.to_string()),
        }
    }

    Ok((session_name.trim().to_string(), urls))
}

pub fn write_session(name: String, urls: Vec<String>) -> io::Result<()> {
    let path = Path::new(".")
        .join("config")
        .join("user")
        .join(format!("{name}.json"));

    if let Some(parent_path) = path.parent() {
        if !parent_path.exists() {
            create_dir_all(parent_path)?;
        }
        let json_record = json!({
            "name": name,
            "urls": urls
        });
        File::create(path)?
            .write_all(to_string_pretty(&json_record)?.as_bytes())?;
    }

    Ok(())
}
